src/Raven.Server/Documents/RemoteAttachmentsStorage.cs

```diff
@@ -374,4 +338,84 @@ public RemoteAttachmentsSender UpdateRemoteAttachmentsFromDatabaseRecord(Databas
+        {
+            using (DocumentIdWorker.GetLoweredIdSliceFromId(context, docId, out var lowerId))
+            using (var doc = Database.DocumentsStorage.Get(context, lowerId, DocumentFields.Data | DocumentFields.Id, throwOnConflict: true))
+            {
+                // remove all entries for processed document
+                foreach (var ticks in allTicks)
+                {
+                    docsTree.MultiDelete(ticks, lowerId);
+                }
+
+                if (doc == null || doc.TryGetMetadata(out var metadata) == false || metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachments) == false)
+                {
+                    // doc was deleted
+                    continue;
+                }
+
+                // lets go over attachments and mark those we proceed as remote
+                foreach (BlittableJsonReaderObject attachmentInMetadata in attachments)
+                {
+                    if (attachmentInMetadata.TryGet(nameof(AttachmentName.RemoteParameters), out BlittableJsonReaderObject remoteParamsObject) == false)
+                        continue;
+
+                    if (remoteParamsObject.TryGet(nameof(RemoteAttachmentParameters.Flags), out object flagsObj) && Enum.TryParse(flagsObj.ToString(), out RemoteAttachmentFlags flags) && flags == RemoteAttachmentFlags.Remote)
+                    {
+                        // this attachment is already remote
+                        continue;
+                    }
+
+                    if (HasPassed(remoteParamsObject, currentTime, MetadataPropertyName) == false)
+                        continue;
+
+                    if (remoteParamsObject.TryGet(nameof(RemoteAttachmentParameters.Identifier), out LazyStringValue identifierFromMetadata) == false)
+                        continue;
+
+                    if (attachmentInMetadata.TryGet(nameof(AttachmentName.Name), out LazyStringValue name) == false ||
+                        attachmentInMetadata.TryGet(nameof(AttachmentName.ContentType), out LazyStringValue contentType) == false ||
+                        attachmentInMetadata.TryGet(nameof(AttachmentName.Hash), out LazyStringValue hash) == false ||
+                        attachmentInMetadata.TryGet(nameof(AttachmentName.Size), out long size) == false)
+                        continue;
+
+                    if (attachmentsToUploadByIdentifier.TryGetValue(identifierFromMetadata, out Dictionary<string, AttachmentRemoteInfo> dic) && dic.TryGetValue(hash, out AttachmentRemoteInfo info))
+                    {
+                        switch (info.Status)
+                        {
+                            case BackgroundWorkInfoStatus.Retry:
+                                // this upload errored lets add back to the tree
+                                remoteParamsObject.TryGet(nameof(RemoteAttachmentParameters.At), out LazyStringValue dateFromMetadata);
```

Comment:
can this be null? I _think_ that `HasPassed` verify that...

Reply:
no, it was checked in `HasPassed`